Step5 - 로또(수동) 리뷰 요청드립니다.

programmer-sjk opened the pull request:
안녕하세요 민수님.

벌써 마지막 리뷰 요청이네요.
항상 밝게 리뷰 해 주셔서 고마웠고 많이 배웠습니다. 감사합니다 😄

programmer-sjk committed:
refactor: 피드백 반영

보너스 볼 매치시, SECOND랑 비교하도록 수정

programmer-sjk committed:
docs(README.md): 요구사항 및 기능 구현 목록 작성

programmer-sjk committed:
feat: 수동으로 구매할 로또 수를 입력받는다

programmer-sjk committed:
feat: 수동으로 로또를 입력받는다

programmer-sjk committed:
feat: 수동, 자동 로또 개수를 출력한다

programmer-sjk committed:
feat: 전체 로또 정보를 출력한다

programmer-sjk committed:
refactor: 테스트 코드 가독성 좋게 변경, lint 수정

programmer-sjk committed:
refactor: Rank 순위 로직 수정

2,3등에 대한 로직을 private 함수로 변경

mskangg commented on README.md:
```diff
@@ -113,4 +113,35 @@
\ No newline at end of file
+- [x] 2등을 출력하는 로직을 추가한다
+
+## 5단계 - 로또(수동)
```
👍 👍 👍

mskangg commented on src/main/java/lotto/domain/LottoMarket.java:
```diff
@@ -2,26 +2,41 @@
 
-    public static List<LottoTicket> sell(Money money, LottoGenerator lottoGenerator) {
-        return make(lottoGenerator, getLottoAmount(money));
+    public static List<LottoTicket> sell(Money money, List<String> manualInputs) {
```
`LottoMarket`에 책임이 많아졌네요!
내부적으로 `Money, LottoGenerator` 상태를 갖고있으므로 충분히 객체로 보여집니다.
해당 클래스를 인스턴스화 할 때 로또가격과 전략을 주입받아 생성하는 방향으로 리팩토링 진행하시는건 어떠신가요?

그리고 대체적으로 static 메서드를 많이 사용하셨는데, 정국님이 생각하시는 특별한 이유가 있으신가요??
static 메서드로 호출하는 클래스와 아닌 클래스를 작성하실 때 어떤 기준으로 나누시는지 궁금합니다!

객체지향 관련된 책 중 `엘레강트 오브젝트` 책에서 나와있는 내용 정리한 블로그 링크 남겨드리겠습니다.
https://jessyt.tistory.com/155

programmer-sjk replied:
내부적으로 `상태를 유지할 필요가 없다고 생각할때` static 메서드로 호출하는 클래스로 작성합니다. (InputView, OutputView)

`LottoMarket`도 처음에는 상태를 유지할 필요가 없다고 생각했는데, 기능이 많아지면서 객체로 보여진다는 민수님 말씀에 동의합니다.

공유해주신 링크의 내용처럼 `정적 메소드 사용을 피하라`는 몰랐던 내용인데 관련 내용 찾아 공부해보겠습니다 😄

mskangg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -6,19 +6,16 @@
         Money money = new Money(InputView.getMoney());
-        List<LottoTicket> tickets = LottoMarket.sell(money, new RandomLottoGenerator());
-        OutputView.printTickets(tickets);
+        Quantity manualQuantity = new Quantity(InputView.getManualLottoCount());
```
구매갯수도 원시값포장 하셨군요 👍

mskangg commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -33,6 +34,19 @@ public boolean contain(LottoNumber number) {
     }
 
+    @Override
+    public boolean equals(Object o) {
```
테스트코드 검증을 위해 equals 메서드를 오버라이딩 하신것으로 보입니다.
이 비교 로직은 로또번호의 순서가 다르면 일치하지 않는 버그가 발생합니다.
(로또번호는 순서에 상관없이 비교되어야합니다.)

세부구현사항의 주석을 살펴보면
```java
// In other words, two lists are defined to be equal if they contain the same elements in the same order.
```
List Collection은 해당 index와 element 두개 모두 동일해야 동등하게 구현되어있습니다.

LottoTicket 테스트에 수동로또 발권시 숫자의 순서를 다르게 입력하면 테스트가 실패합니다.

programmer-sjk replied:
사실 처음 PR description 작성할 때 아래와 같이 작성했는데요
아래의 이유들로 괜찮을거라 생각했던 제 자신을 반성하게 됩니다. 크흠...
민수님이 말하는 로또 번호는 순서에 상관없이 비교되어야 한다가 맞다고 생각해요
1. 테스트 코드에서만 사용된다.
2. 1인 개발자인 저는 테스트에서 로또를 비교할 때 1,2,3,4,5,6 정렬된 값을 넣는 것이 당연하다고 생각했다.
3. 로또 피드백 강의에서 자바지기님이 TDD 영상에서 저렇게 하셔서 보통 이렇게 비교하는 구나.. 착각함...
------------------
**처음 PR**
이번에 조금 물어보고 싶은 것은, 테스트 코드에서 로또가 같은지 비교하기 위해
`LottoTicket`에서 equals 메소드를 오버라이딩 했는데요.
`LottoNumber`와 같이 단일값 비교 외에도 저렇게 List 요소가 같은지 비교하는 방법이 괜찮을지 궁금합니다.

mskangg reviewed (changes requested):
마지막 미션 고생하셨습니다!
다만, 바로 머지하기엔 몇가지 피드백 드리고 싶은 부분이 있어 드렸어요!
확인하시고 반영 부탁드립니다! 👍 

> 항상 밝게 리뷰 해 주셔서 고마웠고 많이 배웠습니다. 감사합니다 😄
좋게 봐주셔서 감사합니다. 🙇‍♂️

programmer-sjk committed:
refactor: 코드 리뷰 반영

로또 티켓 equals 메소드와 마켓 객체화
- constant 패키지에 LOTTO.PRICE 추가
- LottoMarket 객체화
- LottoTicket equals 일치 조건 수정

mskangg commented on src/test/java/lotto/domain/LottoTicketTest.java:
```diff
@@ -48,4 +48,13 @@ void lotto_number_contain_test() {
+        LottoTicket compare = new ManualLottoGenerator(Arrays.asList(6, 5, 4, 3, 2, 1)).create();
+
+        assertThat(origin.equals(compare)).isTrue();
+    }
```
미션의 내용과는 별개지만 참고하시라고 코멘트 남깁니다.

![image](https://user-images.githubusercontent.com/27044096/199670020-addc9d21-bda1-425d-8c90-c4386da9692f.png)

newline을 넣어주시면 좋을 것 같아요!
Github에서는 newline을 남기지 않으면 위와 같이 표식이 남습니다.

POSIX 표준에 맞지 않기도 하고 작업시 newline을 붙이는데, 다른 작업자가 붙이지 않을 경우 불필요한 file change가 생기기도 합니다.

저 같은 경우 IntelliJ에서 newline을 자동으로 추가하는 설정을 사용하고 있습니다.

![image](https://user-images.githubusercontent.com/27044096/199670120-51f0f87c-9fdb-4239-8013-c9411ea0be59.png)

`Ensure every saved file ends with a line break`

mskangg reviewed (approved):
피드백 잘 반영해주셨군요 🙏🏻
마지막 미션까지 정말정말 고생 많으셨습니다~~
처음 과정이라 열정이 넘치고 그럴텐데, 총 9주 과정동안 이 마음 변치마시고 페이스조절 잘 하셔서 의미있는 과정이 되셨으면 좋겠습니다!

마지막으로 소소한팁 남겨드렸으니 확인해주세요!
그럼 2주차 JPA 미션도 화이팅!! 🚀
